File: proj.android_studio/ee-x-facebook/src/main/java/com/ee/facebook/FacebookLoginDelegate.java
package com.ee.facebook;

import androidx.annotation.NonNull;

import com.ee.core.IMessageBridge;
import com.ee.core.internal.Thread;
import com.facebook.FacebookCallback;
import com.facebook.FacebookException;
import com.facebook.login.LoginResult;

/**
 * Created by eps on 3/21/18.
 */

class FacebookLoginDelegate implements FacebookCallback<LoginResult> {
    private final IMessageBridge _bridge;
    private final int _tag;

    FacebookLoginDelegate(@NonNull IMessageBridge bridge, int tag) {
        _bridge = bridge;
        _tag = tag;
    }

    @NonNull
    private String k__onSuccess() {
        return "FacebookLoginDelegate_onSuccess_" + _tag;
    }

    @NonNull
    private String k__onFailure() {
        return "FacebookLoginDelegate_onFailure_" + _tag;
    }

    @NonNull
    private String k__onCancel() {
        return "FacebookLoginDelegate_onCancel_" + _tag;
    }

    @Override
    public void onSuccess(LoginResult result) {
        Thread.checkMainThread();
        _bridge.callCpp(k__onSuccess(), Facebook.convertAccessTokenToString(result.getAccessToken()));
    }

    @Override
    public void onError(FacebookException error) {
        Thread.checkMainThread();
        _bridge.callCpp(k__onFailure(), error.getLocalizedMessage());
    }

    @Override
    public void onCancel() {
        Thread.checkMainThread();
        _bridge.callCpp(k__onCancel());
    }
}

File: proj.android_studio/ee-x-facebook/src/main/java/com/ee/facebook/FacebookLoginDelegate.kt
package com.ee.facebook

import androidx.annotation.AnyThread
import com.ee.core.IMessageBridge
import com.ee.core.internal.Thread
import com.facebook.FacebookCallback
import com.facebook.FacebookException
import com.facebook.login.LoginResult
import kotlinx.serialization.ImplicitReflectionSerializer
import kotlinx.serialization.UnstableDefault

/**
 * Created by eps on 3/21/18.
 */
@ImplicitReflectionSerializer
@UnstableDefault
internal class FacebookLoginDelegate(
    private val _bridge: IMessageBridge,
    private val _tag: Int) : FacebookCallback<LoginResult> {
    private val k__onSuccess: String
        @AnyThread get() = "FacebookLoginDelegate_onSuccess_$_tag"

    private val k__onFailure: String
        @AnyThread get() = "FacebookLoginDelegate_onFailure_$_tag"

    private val k__onCancel: String
        @AnyThread get() = "FacebookLoginDelegate_onCancel_$_tag"

    override fun onSuccess(result: LoginResult) {
        Thread.checkMainThread()
        _bridge.callCpp(k__onSuccess, Facebook.convertAccessTokenToString(result.accessToken))
    }

    override fun onError(error: FacebookException) {
        Thread.checkMainThread()
        _bridge.callCpp(k__onFailure, error.localizedMessage ?: "")
    }

    override fun onCancel() {
        Thread.checkMainThread()
        _bridge.callCpp(k__onCancel)
    }
}